### ISSUE TITLE
macos13: presets menu doesn't get focus for mouse actions

### ISSUE THREAD

**Issue by MStraeten** [darktable-org/darktable#12727]
originally reported with #12708 

when opening the presets menu of a module the menu is displayed but doesn't get focus for mouse actions
if mouse pointer is moved over slowly one menu item the tooltip of the module parameter just under the menu is shown (the mouse pointer was over JPG 1024 item):
![image](https://user-images.githubusercontent.com/39386816/198697113-ead86f2c-9aec-40d1-bd6c-b39e6b4074ad.png)

after the tooltip disappears when moving the mouse further the menu gets focus and the items are highlighted on mouse over and can be selected.

selection of menu items using keyboard is possible but just for shared utility modules like export, tagging, ... and module groups or module order in darkroom then a selection is possible.
for menuitems of processing modules in darkroom right panel keyboards selection doesn't work

**To Reproduce**
just reproducible with macos13 ventura
1. Go to lighttable
2. expand export module
3. open presets menu
4. select an item --> nothing happens
5. like 4 but hover mouse over an item having an parameter below the menu and wait for the tooltip (you can also move the mouse pointer over a module header to get that tooltip
6. move mouse back so th tooltip disappears
7. then you can successfully highlight and select an item

**Platform**
* darktable version : m1 and x86_64, reported for 4.0.1 but also reproduced with a couple of 4.1 builds using sdk 12.3 and sdk 13  builds
* OS                : macOS13 Ventura

**Comment by MStraeten** [comment]
seems to be gtk issue: [https://gitlab.gnome.org/GNOME/gtk/-/issues/5305]
remaining question for a workaround: why does the behaviour of preset menus for utility modules differ from that in the processing modules for keyboard actions?

**Comment by pradyut-c** [comment]
I am also facing same issues in mac os 
I tried ARM and intel version both on my m1
But presets menus not selecting properly on both versions..

darktable team please fix this issue it’s breaking the workflow please 🙏 🙏

**Comment by MStraeten** [comment]
It’s a gtk issue the darktable team cannot fix. Unfortunately breaking compatibility is a quite common side effect of apple’s macOS update process. They simply don’t care about third party software, so you’d better check for side effects before upgrading an production system.
BTW: no need to collect confirmations of the issue here.

**Comment by parafin** [comment]
Yeah, I'm not sure we can workaround this issue on darktable's side. I'm not going to patch GTK, so we just have to wait for their release fixing this bug.

**Comment by CommunyAS** [comment]
Having the same issue - and, additionally problems to duplicate or delete modules - I found workarounds in [#12708](https://github.com/darktable-org/darktable/issues/12708):

https://github.com/darktable-org/darktable/issues/12708#issuecomment-1293619295

https://github.com/darktable-org/darktable/issues/12708#issuecomment-1295196780 - this one by Mitraten

**Comment by MStraeten** [comment]
just checked it - the fix works fine for me (having gtk3 @3.24.34_2+quartz) but i didn't checked the whole application yet.
how to apply it (on your own risk):

then prepare gtk3 patch
```
mkdir -p ~/ports/gnome/gtk3/files 
cp -R "$(port dir gtk3)" ~/ports/gnome
curl -Lo ~/ports/gnome/gtk3/files/fix-some-ventura-mouse-problems.patch https://gitlab.gnome.org/Infrastructure/gimp-macos-build/-/raw/c99f829a62b7e5b60616d8d977b08169927842cf/ports/gnome/gtk3/files/fix-some-ventura-mouse-problems.patch
```
then patch Portfile (after checking you are on gtk 3.24.34):
`patch ~/ports/gnome/gtk3/Portfile < `[patchPortfile.txt](https://github.com/darktable-org/darktable/files/10059324/patchPortfile.txt)
this increases version to 3 (gtk3 @3.24.34_3+quartz) to be able to revert to already installed version 2 if the upgrade fails ;)

then update ports `portindex ~/ports` and  `sudo port selfupdate`
do all updates `sudo port upgrade outdated` and rebuild darktable

**since that patch isn't approved and merged yet, don't use it if you need a reliable gtk3 library**

**Comment by Nilvus** [comment]
@MStraeten: as this issue and #12708 seems more related to a Gtk issue than darktable one, maybe we should close them? Could you confirm?

**Comment by parafin** [comment]
We should close one of these and let the other be open. No need for two tickets, but open ones are more likely to be found, so it will cause less duplicates to be reported.

**Comment by MStraeten** [comment]
Both issue describes the same issue, this is a bit more specific.
and has a workaround for custom builds until gtk3 is fixed ;)

**Comment by MStraeten** [comment]
unfortunately the latest patches of gtk3 for this issue requires gtk3 2.24.35 but macports currently is on 3.24.34 
so keep your fingers crossed for 4.2

**Comment by MStraeten** [comment]
Latest patch was merged yet: https://gitlab.gnome.org/Infrastructure/gimp-macos-build/-/tree/master/ports/gnome/gtk3

**Comment by Tuxie** [comment]
I downloaded the 4.2 macOS build from https://github.com/darktable-org/darktable/releases/tag/release-4.2.0
They still have this problem, and they also seem to be built with gtk+ 3.24.34 instead of 3.24.35

**Comment by parafin** [comment]
3.24.35 doesn’t has this bug fixed anyway, so it doesn’t matter. And I won’t maintain any more custom patches, I just got rid of last one recently. This bug will be fixed when GTK fixes it, so complain to them and Apple.

**Comment by MStraeten** [comment]
unless macports integrates 3.24.35 and the patch there’s just a working solution (see earlier) post working with 3.24.34 - but that’s not the final solution, so a package maintainer having a stable release in mind can’t base on it.
So it’s up to you to decide, if a not maintained build using that patch is ok for you. If so, you can try https://discuss.pixls.us/t/info-macos-13-breaks-darktable/33267/8

**Comment by Tuxie** [comment]
GTK+ 3.24.36 was released December 22nd. Its changelog contains "MacOS: Use NSTrackingArea".

https://ftp.acc.umu.se/pub/GNOME/sources/gtk+/3.24/gtk+-3.24.36.news

**Comment by parafin** [comment]
Please let me know once macports update their version (https://ports.macports.org/port/gtk3/), we'll decide how to proceed after that.

**Comment by zisoft** [comment]
I build darktable on my Mac using homebrew. Homebrew has already updated GTK+ to 3.24.36 and I can confirm that this bug is resolved.

**Comment by Nilvus** [comment]
@MStraeten: maybe that could be closed now if this is fixed on Homebrew?

**Comment by Tuxie** [comment]
@Nilvus It's not in MacPorts yet, and consequently also not in the official Darktable packages for macOS.
https://trac.macports.org/ticket/66487

**Comment by MStraeten** [comment]
please keep it open until a fixed official package exists - since users that migrated to ventura or got a new mac might report this issue it doesn’t make sense to explain the whole shebang with new issue reports ;)

**Comment by Nilvus** [comment]
Ok, no problem. Was just in case. Thanks for your replies.

**Comment by MStraeten** [comment]
was able to build also on arm64, with following receipt installing gtk3 3.24.36 via gtk3-devel port (in case it doesn't work it at least doesn't damage your existing gtk3 port )
so this can be used for intel and M1 macs

prerequisite: macports environment following https://github.com/darktable-org/darktable/blob/master/packaging/macosx/BUILD-ARM64.txt or https://github.com/darktable-org/darktable/blob/master/packaging/macosx/BUILD.txt exists and dependencies are installed.
 
```
mkdir -p ~/ports/gnome/gtk3-devel/files ~/ports/gnome
cp -R "$(port dir gtk3-devel)" ~/ports/gnome
curl -o ~/ports/gnome/gtk3-devel/patches.zip https://trac.macports.org/zip-attachment/ticket/66487
unzip  -d ~/ports/gnome/gtk3-devel/ ~/ports/gnome/gtk3-devel/patches.zip
mv ~/ports/gnome/gtk3-devel/patch-meson_build.diff ~/ports/gnome/gtk3-devel/files/patch-meson_build.diff
patch ~/ports/gnome/gtk3-devel/Portfile < ~/ports/gnome/gtk3-devel/patch-gtk3-devel_Portfile_v0.1.diff

portindex ~/ports
```
following guides above gtk3 is installed and activated so you need to deactivate it before installing gtk3-devel
```
sudo port deactivate gtk3
sudo port install gtk3-devel
```



then you're ready to build darktable and get rid of the menu issue ...

**Comment by MStraeten** [comment]
It seems the gtk3.24.36 package built with that receipt breaks display profile handling. See:
https://discuss.pixls.us/t/gamma-saturation-difference-between-dt-and-other-apps-on-mac/35038/11
Can someone with a homebrew build using gtk3.24.36 check, if that reported behaviour also occurs with their builds?

**Comment by parafin** [comment]
Could it be more about cairo version than gtk? As far as I remember color profile stuff happens mainly in cairo.

**Comment by MStraeten** [comment]
latest cairoupdate in macports that is used in my macports build was 24.12.2021 - it was recently updated (8hrs ago); i need to update if i'm back at my macos machine.
So the main difference between my latest build with proper handling (6.1.2023) and broken handling (10.1.2023) was replacing gtk3 3.24.34 by gtk3 3.24.36 using the gtk3-devel portfile from https://trac.macports.org/ticket/66487 to update that to 3.24.36.  Both built with the same cairo library.

**Comment by MStraeten** [comment]
got a homebrew build - and it shows the same behaviour (https://discuss.pixls.us/t/gamma-saturation-difference-between-dt-and-other-apps-on-mac/35038/20) . so it's an upstream issue with gtk3 3.24.36

possible workaround: use a custom display profile copied into ~/.config/darktable/color/out

**Comment by parafin** [comment]
There is actually a commented out code in osx.mm that gets monitor profile. But I’m not sure if it still works. In any case I suppose this is smth that’s going to be reverted in gtk, so better not workaround this in dt.

**Comment by MStraeten** [comment]
Got response from gtk3 team:
[John Ralls](https://gitlab.gnome.org/jralls) [commented](https://gitlab.gnome.org/GNOME/gtk/-/issues/5579#note_1661328):

The big difference between 3.24.34 and 3.24.36 is the change from drawing with Cairo's Quartz backend that uses CoreGraphics drawing primitives to drawing with Cairo's bitma@p backend and copying the image to a CoreAnimation layer. Part of that changes the function used to get the colorspace for the CGContext from CGColorSpaceCreateDeviceRGB() to CGDisplayCopyColorSpace (CGMainDisplayID ()) in order to prevent Quartz from very expensively converting the colorspace when it transfers the layer to the GPU for display. I made those changes to improve performance, see [#3714](https://gitlab.gnome.org/GNOME/gtk/-/issues/3714).

[@MStraeten](https://gitlab.gnome.org/MStraeten) can you find out from the Darktable devs where the colorspace management code is, particularly any that specializes handling on macOS?

**Comment by parafin** [comment]
It's as if GTK people are actively trying to tell developers to piss off and stop using GTK for anything serious.

**Comment by MStraeten** [comment]
https://gitlab.gnome.org/GNOME/gtk/-/issues/5579 was fixed

to integrate that fix into the gtk3 3.24.36 macports description (https://github.com/darktable-org/darktable/issues/12727#issuecomment-1407005092) 
add [CMPatch.diff.txt](https://github.com/darktable-org/darktable/files/10714838/CMPatch.diff.txt) into ports/gnome/gtk3-devel/files (ditch the .txt extension) 
add `patchfiles-append CMPatch.diff` into the generated Portfile (e.g. line 63)

**Comment by parafin** [comment]
Did you test it yourself and confirm that it’s indeed fixed? Just checking to be sure.

This whole approach “we don’t promise anything regarding colors” should be a red flag for anyone thinking to create an application working with images using GTK. Which is ironic, given its roots in GIMP (GIMP ToolKit being GTK). Sadly we’re way too far down the rabbit hole at this point.

**Comment by MStraeten** [comment]
yes. i integrated the patch, then did processing with gtk3 3.24.34, gtk3 3.24.36 original and gtk3 3.24.36 patched
the display of image with 'system display profile' is identical with 3.24.34 and 3.24.36 patched - the original gtk3 3.24.36 show the difference.

builds for arm64 and x86_64 of recent darktable-4.2.x branch are in progress ...

**Comment by maxwxyz** [comment]
@MStraeten is this fix included in the new release 4.2.1 or are there now 2 versions?

**Comment by parafin** [comment]
That is a question for me, and yes, this fix is included in 4.2.1.

**Comment by MStraeten** [comment]
then we can close this here ...

**Comment by MStraeten** [comment]
update for gtk3 3.24.37 including the colormanagement patch:

```
cp -R "$(port dir gtk3-devel)" ~/ports/gnome
curl -o ~/ports/gnome/gtk3-devel/patches.zip https://trac.macports.org/zip-attachment/ticket/66487
unzip  -d ~/ports/gnome/gtk3-devel/ ~/ports/gnome/gtk3-devel/patches.zip
mv ~/ports/gnome/gtk3-devel/patch-meson_build.diff ~/ports/gnome/gtk3-devel/files/patch-meson_build.diff
patch ~/ports/gnome/gtk3-devel/Portfile < ~/ports/gnome/gtk3-devel/patch-gtk3-devel_Portfile_v0.2.diff
```
```
sudo port deactivate gtk3
sudo port install gtk3-devel
```

**Comment by zisoft** [comment]
Homebrew is updated to gtk3 3.24.37 as well.

**Comment by MStraeten** [comment]
macports update for gtk3 3.24.37 is in the pipe 
https://github.com/macports/macports-ports/pull/18385
